File: .repo/system-design-mastery-module-2-database-fundamentals/3-sharding-strategies/1-java/src/main/java/academy/starci/sharding/users/UsersController.java
package academy.starci.sharding.users;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api")
public class UsersController {

    private final UsersService usersService;

    public UsersController(UsersService usersService) {
        this.usersService = usersService;
    }

    @PostMapping("/users")
    public ResponseEntity<UserDocument> createUser(@RequestBody UserDocument user) {
        UserDocument created = usersService.create(user);
        return ResponseEntity.status(HttpStatus.CREATED).body(created);
    }

    @GetMapping("/users/{userId}")
    public ResponseEntity<UserDocument> findByUserId(@PathVariable("userId") String userId) {
        return usersService.findByUserId(userId)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping("/users/country/{country}")
    public ResponseEntity<List<UserDocument>> findByCountry(@PathVariable("country") String country) {
        List<UserDocument> users = usersService.findByCountry(country);
        return ResponseEntity.ok(users);
    }

    @GetMapping("/users/tier/{tier}")
    public ResponseEntity<List<UserDocument>> findByTier(@PathVariable("tier") String tier) {
        List<UserDocument> users = usersService.findByTier(tier);
        return ResponseEntity.ok(users);
    }

    @PostMapping("/seed/{count}")
    public ResponseEntity<Map<String, Object>> seed(@PathVariable("count") int count) {
        Map<String, Object> result = usersService.seed(count);
        return ResponseEntity.status(HttpStatus.CREATED).body(result);
    }

    @GetMapping("/shards/distribution")
    public ResponseEntity<Map<String, Object>> getShardDistribution() {
        Map<String, Object> result = usersService.getShardDistribution();
        return ResponseEntity.ok(result);
    }

    @GetMapping("/shards/status")
    public ResponseEntity<Map<String, Object>> getShardStatus() {
        Map<String, Object> result = usersService.getShardStatus();
        return ResponseEntity.ok(result);
    }

    @GetMapping("/shards/explain")
    public ResponseEntity<Map<String, Object>> explainQuery(
            @RequestParam("field") String field,
            @RequestParam("value") String value) {
        Map<String, Object> result = usersService.explainQuery(field, value);
        return ResponseEntity.ok(result);
    }
}

File: .repo/system-design-mastery-module-2-database-fundamentals/3-sharding-strategies/2-csharp/Controllers/UsersController.cs
using Microsoft.AspNetCore.Mvc;
using ShardingStrategies.Models;
using ShardingStrategies.Services;
using System.Threading.Tasks;

namespace ShardingStrategies.Controllers
{
    [ApiController]
    [Route("api")]
    public class UsersController : ControllerBase
    {
        private readonly UsersService _usersService;

        public UsersController(UsersService usersService)
        {
            _usersService = usersService;
        }

        [HttpPost("users")]
        public async Task<IActionResult> CreateUser([FromBody] UserEntity user)
        {
            var created = await _usersService.CreateAsync(user);
            return StatusCode(201, created);
        }

        [HttpGet("users/{userId}")]
        public async Task<IActionResult> FindByUserId([FromRoute] string userId)
        {
            var user = await _usersService.FindByUserIdAsync(userId);
            if (user == null)
            {
                return NotFound();
            }
            return Ok(user);
        }

        [HttpGet("users/country/{country}")]
        public async Task<IActionResult> FindByCountry([FromRoute] string country)
        {
            var users = await _usersService.FindByCountryAsync(country);
            return Ok(users);
        }

        [HttpGet("users/tier/{tier}")]
        public async Task<IActionResult> FindByTier([FromRoute] string tier)
        {
            var users = await _usersService.FindByTierAsync(tier);
            return Ok(users);
        }

        [HttpPost("seed/{count}")]
        public async Task<IActionResult> Seed([FromRoute] int count)
        {
            var result = await _usersService.SeedAsync(count);
            return StatusCode(201, result);
        }

        [HttpGet("shards/distribution")]
        public async Task<IActionResult> GetShardDistribution()
        {
            var result = await _usersService.GetShardDistributionAsync();
            return Ok(result);
        }

        [HttpGet("shards/status")]
        public async Task<IActionResult> GetShardStatus()
        {
            var result = await _usersService.GetShardStatusAsync();
            return Ok(result);
        }

        [HttpGet("shards/explain")]
        public async Task<IActionResult> ExplainQuery([FromQuery] string field, [FromQuery] string value)
        {
            var result = await _usersService.ExplainQueryAsync(field, value);
            return Ok(result);
        }
    }
}
